File: genkit_backend/lib/flows/focus_insight_flow.dart
// ─────────────────────────────────────────────────────────────────────────────
// genkit_backend/lib/flows/focus_insight_flow.dart
//
// FIX: Extracted from task_generator_flow.dart — this function was previously
//      in that file but main.dart tried to import it from this non-existent
//      file, causing a compile error.
//
// Genkit flow: focusInsight
// Called after a Pomodoro/focus session ends.
// Returns an encouraging insight, a next-step hint, and a coin reward.
// ─────────────────────────────────────────────────────────────────────────────

import 'package:genkit/genkit.dart';
import 'package:genkit_google_genai/genkit_google_genai.dart';

GenkitFlow registerFocusInsightFlow(Genkit ai) {
  return ai.defineFlow(
    name: 'focusInsight',
    inputSchema: z.object({
      'taskTitle':       z.string(),
      'goalTitle':       z.string(),
      'durationMinutes': z.number().int(),
      'completed':       z.boolean(),
    }),
    outputSchema: z.object({
      'insight':      z.string(),
      'nextStepHint': z.string(),
      'coinsEarned':  z.number().int(),
    }),
    fn: (input) async {
      final taskTitle = input['taskTitle'] as String;
      final goalTitle = input['goalTitle'] as String;
      final duration  = (input['durationMinutes'] as num).toInt();
      final completed = input['completed'] as bool;

      // ENHANCE: Coin range is calculated here and passed as a hint, so the
      // model can suggest an appropriate value rather than always picking
      // exactly coinBase.  The backend still owns the final value.
      final coinBase  = completed
          ? (duration * 0.8).round()
          : (duration * 0.4).round();
      final coinMin   = (coinBase - 5).clamp(1, 9999);
      final coinMax   = coinBase + 10;
      final status    = completed ? 'fully completed' : 'partially completed';

      final prompt = '''
A Goal Digger user just $status a focus session.
Task: "$taskTitle"  |  Goal: "$goalTitle"  |  Duration: $duration minutes.

Write:
1. "insight": A brief, specific, encouraging observation about this session (1 sentence).
   Reference the task or goal by name — don't be generic.
2. "nextStepHint": A concrete hint for what to do next to keep momentum (1 sentence).
3. "coinsEarned": Integer coin reward between $coinMin and $coinMax.
   Completed sessions earn more. Longer sessions earn more within the range.

Respond ONLY with valid JSON:
{
  "insight": "...",
  "nextStepHint": "...",
  "coinsEarned": $coinBase
}
''';

      final response = await ai.generate(
        model: googleGenAI('gemma-4-it'),
        prompt: prompt,
        config: GenerationConfig(
          temperature: 0.7,
          maxOutputTokens: 256,
          responseMimeType: 'application/json',
        ),
      );

      final parsed = response.outputAsMap();

      // Guard coin value to stay within the intended range
      final rawCoins = (parsed['coinsEarned'] as num?)?.toInt() ?? coinBase;
      final coinsEarned = rawCoins.clamp(coinMin, coinMax);

      return {
        'insight':      parsed['insight']      ?? '',
        'nextStepHint': parsed['nextStepHint'] ?? '',
        'coinsEarned':  coinsEarned,
      };
    },
  );
}

File: functions/src/flows/focusInsightFlow.ts
// functions/src/flows/focusInsightFlow.ts
//
// Generates a post-session insight after the user completes (or stops) a
// focus timer. Rewards completed sessions with coins and an AI reflection.

import { z } from "genkit";
import { getAI, defaultModel } from "../ai";
import { parseModelJson } from "../json";

const FocusInsightInputSchema = z.object({
  taskTitle:       z.string(),
  goalTitle:       z.string(),
  durationMinutes: z.number().min(1),
  completed:       z.boolean(),
});

const FocusInsightOutputSchema = z.object({
  insight:      z.string(),
  coinsEarned:  z.number().default(0),
  badge:        z.string().default(""),
  nextStepHint: z.string().default(""),  // added to match Flutter FocusInsightResponse
});

export function defineFocusInsightFlow() {
  const ai = getAI();

  return ai.defineFlow(
    {
      name:         "focusInsight",
      inputSchema:  FocusInsightInputSchema,
      outputSchema: FocusInsightOutputSchema,
    },
    async (input) => {
      // Deterministic coin reward — not AI-controlled so it can't be gamed
      const coinsEarned = input.completed
        ? Math.min(50, Math.max(5, Math.round(input.durationMinutes / 2)))
        : 0;

      const badge =
        input.completed && input.durationMinutes >= 45
          ? "🏅 Deep Work"
          : input.completed && input.durationMinutes >= 25
          ? "⚡ Flow State"
          : input.completed
          ? "✅ Session Done"
          : "";

      const prompt = `
You are a focus coach in a productivity app. The user just ${
        input.completed ? "completed" : "stopped early"
      } a ${input.durationMinutes}-minute session.

Task: "${input.taskTitle}"
Goal: "${input.goalTitle}"
Completed: ${input.completed}
${input.completed ? `Coins earned: ${coinsEarned}` : ""}

Write ONE sentence (max 20 words) that:
- If completed: celebrates the win and connects it to the bigger goal
- If stopped early: is non-judgmental and motivates them to try again

Respond ONLY with valid JSON: { "insight": "..." }`.trim();

      try {
        const { text } = await ai.generate({
          model: defaultModel,
          prompt,
          config: {
            temperature: 0.6,
            maxOutputTokens: 80,
            responseMimeType: "application/json",
          },
        });

        const parsed = parseModelJson<{ insight: string }>(text);
        const insight = parsed.insight?.trim();

        if (insight) {
          const nextStepHint = input.completed
            ? `Keep the momentum — start your next task within 10 minutes.`
            : `When you're ready, try again with a shorter 10-minute session.`;
          return { insight, coinsEarned, badge, nextStepHint };
        }
      } catch (e) {
        // Fall through to static message
      }

      const staticInsight = input.completed
        ? `Great work — ${input.durationMinutes} focused minutes brings "${input.goalTitle}" closer to done.`
        : `Even a short session counts. Jump back in when you're ready.`;

      const staticNextStep = input.completed
        ? `Review what you just completed and pick the next action.`
        : `Take a short break, then try again with a smaller time block.`;

      return { insight: staticInsight, coinsEarned, badge, nextStepHint: staticNextStep };
    }
  );
}
